cmd/nvidia-cdi-hook/README.md

```diff
@@ -0,0 +1,14 @@
+# NVIDIA CDI Hook
+
+The NVIDIA Container Device Interface (CDI) Hook CLI `nvidia-cdi-hook` provides
+the runtime hook capabilities called from a CDI file.
+
+## Functionality
+
+The `nvidia-cdi-hook` CLI provides the following functionality:
+
+* `chmod` - Change the permissions of a file or directory inside the directory path to be mounted into a container.
+* `create-symlinks` - Create symlinks inside the directory path to be mounted into a container.
+* `update-ldcache` - Update the dynamic linker cache inside the directory path to be mounted into a container.
+
+The CDI itself is created separately, normally using the `nvidia-ctk cdi` command.
```

Comment:
Can you document how to actually use the hook? Is it triggered automatically or users need to configure something?

Reply:
I don't quite understand. It doesn't change from current behaviour. It's inserted into the cdi yaml when you run `nvidia-ctk cdi generate`

Reply:
So, can you document the current behavior?

For me it does not work by default though, `nvidia-ctk cdi generate` creates a CDI spec where is used only in `hooks:` sections, but it is missing in the `mounts:` section. Then e.g. when I run `podman run --rm -it --gpus all nvidia/cuda:12.4.1-devel-ubuntu22.04`, `/usr/bin/nvidia-ctk` is not available in the container.

Reply:
> So, can you document the current behavior?

Sure, on its way.

Reply:
> For me it does not work by default though, nvidia-ctk cdi generate creates a CDI spec where nvidia-ctk is used only in hooks: sections, but it is missing in the mounts: section. Then e.g. when I run podman run --rm -it --gpus all nvidia/cuda:12.4.1-devel-ubuntu22.04, /usr/bin/nvidia-ctk is not available in the container.

Do you mind posting your CDI here? I have 2 Jetsons, but unfortunately no access to them for the next week, so I have nothing ti generate CDIs on, in order to tests it. I also don't have remote access to anything with proper GPUs, so I am stuck working this through here without proper testing.

Reply:
> So, can you document the current behavior?

Done.

Reply:
> It shouldn't be. containerd (or whatever CDI-compliant high-level runtime you use) is supposed to run the various hooks before launching the container, at least as far as I understand it.

Oh, that clarifies it. I've checked properly and e.g. the `create-symlinks` and `update-ldcache` hooks were indeed run, the container has the relevant symbolic links.

Reply:
> Oh, that clarifies it. 

Cool, I am glad it works.

Do you mind sharing your generated CDI yaml? It will give me a better basis to check the work?

Reply:
Sure, here it is: https://gist.github.com/lahwaacz/0beed267b76ddb1c6d634e5ca2a3a229

Reply:
Thank you. Heh, it is so small, compared to when I ran it on my Jetson Xavier NX. Last run had 3367 lines 😬 

I see it didn't include `nvidia-cdi-hook` for any of them, but rather `nvidia-ctk`. Is this generated by `nvidia-ctk` from the latest commit in this PR? If so, I definitely missed something.

Reply:
Hmm, the code looks right. Do you mind pulling latest commit down, recompiling, and running `nvidia-ctk cdi generate` again?

Reply:
Also, do you know of any way to emulate GPUs in a VM or container, so I can run this without accessing an actual device?

Reply:
Err, sorry - that was generated with the 1.15.0 version. This one is generated with the code from your branch: https://gist.github.com/lahwaacz/e28cf9aab89cc5175eee0c57436acc4d

Note there is a difference especially in the last `devices:` section (I don't know if it may be related to the 1.15.0 version existing in the system as well). There were some warnings when I run the command:

```
$ ./build/nvidia-ctk cdi generate
WARN[0000] Failed to locate nvidia-cdi-hook: pattern nvidia-cdi-hook not found
INFO[0000] Auto-detected mode as "nvml"
INFO[0000] Selecting /dev/nvidia0 as /dev/nvidia0
WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--card; ignoring
WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--render; ignoring
INFO[0000] Using driver version 550.67
INFO[0000] Selecting /dev/nvidia-modeset as /dev/nvidia-modeset
INFO[0000] Selecting /dev/nvidia-uvm-tools as /dev/nvidia-uvm-tools
INFO[0000] Selecting /dev/nvidia-uvm as /dev/nvidia-uvm
INFO[0000] Selecting /dev/nvidiactl as /dev/nvidiactl
Vim: Reading from stdin...
INFO[0000] Selecting /usr/lib/libnvidia-egl-gbm.so.1.1.1 as /usr/lib/libnvidia-egl-gbm.so.1.1.1
INFO[0000] Selecting /usr/share/glvnd/egl_vendor.d/10_nvidia.json as /usr/share/glvnd/egl_vendor.d/10_nvidia.json
INFO[0000] Selecting /usr/share/vulkan/icd.d/nvidia_icd.json as /usr/share/vulkan/icd.d/nvidia_icd.json
WARN[0000] Could not locate vulkan/icd.d/nvidia_layers.json: pattern vulkan/icd.d/nvidia_layers.json not found
INFO[0000] Selecting /usr/share/vulkan/implicit_layer.d/nvidia_layers.json as /usr/share/vulkan/implicit_layer.d/nvidia_layers.json
INFO[0000] Selecting /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json as /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json
INFO[0000] Selecting /usr/share/egl/egl_external_platform.d/10_nvidia_wayland.json as /usr/share/egl/egl_external_platform.d/10_nvidia_wayland.json
INFO[0000] Selecting /usr/share/nvidia/nvoptix.bin as /usr/share/nvidia/nvoptix.bin
WARN[0000] Could not locate nvidia/xorg/nvidia_drv.so: pattern nvidia/xorg/nvidia_drv.so not found
INFO[0000] Selecting /usr/lib/nvidia/xorg/libglxserver_nvidia.so.550.67 as /usr/lib/nvidia/xorg/libglxserver_nvidia.so.550.67
WARN[0000] Could not locate X11/xorg.conf.d/10-nvidia.conf: pattern X11/xorg.conf.d/10-nvidia.conf not found
INFO[0000] Selecting /usr/lib/libEGL_nvidia.so.550.67 as /usr/lib/libEGL_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLESv1_CM_nvidia.so.550.67 as /usr/lib/libGLESv1_CM_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLESv2_nvidia.so.550.67 as /usr/lib/libGLESv2_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLX_nvidia.so.550.67 as /usr/lib/libGLX_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libcuda.so.550.67 as /usr/lib/libcuda.so.550.67
INFO[0000] Selecting /usr/lib/libcudadebugger.so.550.67 as /usr/lib/libcudadebugger.so.550.67
INFO[0000] Selecting /usr/lib/libnvcuvid.so.550.67 as /usr/lib/libnvcuvid.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-allocator.so.550.67 as /usr/lib/libnvidia-allocator.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-cfg.so.550.67 as /usr/lib/libnvidia-cfg.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-eglcore.so.550.67 as /usr/lib/libnvidia-eglcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-encode.so.550.67 as /usr/lib/libnvidia-encode.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-fbc.so.550.67 as /usr/lib/libnvidia-fbc.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glcore.so.550.67 as /usr/lib/libnvidia-glcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glsi.so.550.67 as /usr/lib/libnvidia-glsi.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glvkspirv.so.550.67 as /usr/lib/libnvidia-glvkspirv.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-gpucomp.so.550.67 as /usr/lib/libnvidia-gpucomp.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ml.so.550.67 as /usr/lib/libnvidia-ml.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ngx.so.550.67 as /usr/lib/libnvidia-ngx.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-nvvm.so.550.67 as /usr/lib/libnvidia-nvvm.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-opticalflow.so.550.67 as /usr/lib/libnvidia-opticalflow.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-pkcs11-openssl3.so.550.67 as /usr/lib/libnvidia-pkcs11-openssl3.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-pkcs11.so.550.67 as /usr/lib/libnvidia-pkcs11.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ptxjitcompiler.so.550.67 as /usr/lib/libnvidia-ptxjitcompiler.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-rtcore.so.550.67 as /usr/lib/libnvidia-rtcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-tls.so.550.67 as /usr/lib/libnvidia-tls.so.550.67
INFO[0000] Selecting /usr/lib/libnvoptix.so.550.67 as /usr/lib/libnvoptix.so.550.67
INFO[0000] Selecting /run/nvidia-persistenced/socket as /run/nvidia-persistenced/socket
WARN[0000] Could not locate /nvidia-fabricmanager/socket: pattern /nvidia-fabricmanager/socket not found
WARN[0000] Could not locate /tmp/nvidia-mps: pattern /tmp/nvidia-mps not found
INFO[0000] Selecting /lib/firmware/nvidia/550.67/gsp_ga10x.bin as /lib/firmware/nvidia/550.67/gsp_ga10x.bin
INFO[0000] Selecting /lib/firmware/nvidia/550.67/gsp_tu10x.bin as /lib/firmware/nvidia/550.67/gsp_tu10x.bin
INFO[0000] Selecting /usr/bin/nvidia-smi as /usr/bin/nvidia-smi
INFO[0000] Selecting /usr/bin/nvidia-debugdump as /usr/bin/nvidia-debugdump
INFO[0000] Selecting /usr/bin/nvidia-persistenced as /usr/bin/nvidia-persistenced
INFO[0000] Selecting /usr/bin/nvidia-cuda-mps-control as /usr/bin/nvidia-cuda-mps-control
INFO[0000] Selecting /usr/bin/nvidia-cuda-mps-server as /usr/bin/nvidia-cuda-mps-server
INFO[0000] Generated CDI spec with version 0.5.0
```

Reply:
Two questions. 

1. Do you mind rerunning with 1.15.0 and posting the logs, so I can compare?
2. Is the format on the second one [here](https://gist.github.com/lahwaacz/e28cf9aab89cc5175eee0c57436acc4d) messed up? The yaml indentation is off; compare with the first one you posted [here](https://gist.github.com/lahwaacz/0beed267b76ddb1c6d634e5ca2a3a229). Is that a copy-paste error? Or is the original yaml off?

Reply:
```
podman run --rm -it --gpus all nvidia/cuda:12.4.1-devel-ubuntu22.04
```
This does not use CDI. The command you want is:
```
podman run --rm -it --device nvidia.com/gpu=all nvidia/cuda:12.4.1-devel-ubuntu22.04
```

Reply:
@deitch The logs with 1.15.0 are:
```
INFO[0000] Auto-detected mode as "nvml"
INFO[0000] Selecting /dev/nvidia0 as /dev/nvidia0
INFO[0000] Selecting /dev/dri/card1 as /dev/dri/card1
INFO[0000] Selecting /dev/dri/renderD128 as /dev/dri/renderD128
INFO[0000] Using driver version 550.67
INFO[0000] Selecting /dev/nvidia-modeset as /dev/nvidia-modeset
INFO[0000] Selecting /dev/nvidia-uvm-tools as /dev/nvidia-uvm-tools
INFO[0000] Selecting /dev/nvidia-uvm as /dev/nvidia-uvm
INFO[0000] Selecting /dev/nvidiactl as /dev/nvidiactl
INFO[0000] Selecting /usr/lib/libnvidia-egl-gbm.so.1.1.1 as /usr/lib/libnvidia-egl-gbm.so.1.1.1
INFO[0000] Selecting /usr/share/glvnd/egl_vendor.d/10_nvidia.json as /usr/share/glvnd/egl_vendor.d/10_nvidia.json
INFO[0000] Selecting /usr/share/vulkan/icd.d/nvidia_icd.json as /usr/share/vulkan/icd.d/nvidia_icd.json
WARN[0000] Could not locate vulkan/icd.d/nvidia_layers.json: pattern vulkan/icd.d/nvidia_layers.json not found
INFO[0000] Selecting /usr/share/vulkan/implicit_layer.d/nvidia_layers.json as /usr/share/vulkan/implicit_layer.d/nvidia_layers.json
INFO[0000] Selecting /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json as /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json
INFO[0000] Selecting /usr/share/egl/egl_external_platform.d/10_nvidia_wayland.json as /usr/share/egl/egl_external_platform.d/10_nvidia_wayland.json
INFO[0000] Selecting /usr/share/nvidia/nvoptix.bin as /usr/share/nvidia/nvoptix.bin
WARN[0000] Could not locate nvidia/xorg/nvidia_drv.so: pattern nvidia/xorg/nvidia_drv.so not found
INFO[0000] Selecting /usr/lib/nvidia/xorg/libglxserver_nvidia.so.550.67 as /usr/lib/nvidia/xorg/libglxserver_nvidia.so.550.67
WARN[0000] Could not locate X11/xorg.conf.d/10-nvidia.conf: pattern X11/xorg.conf.d/10-nvidia.conf not found
INFO[0000] Selecting /usr/lib/libEGL_nvidia.so.550.67 as /usr/lib/libEGL_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLESv1_CM_nvidia.so.550.67 as /usr/lib/libGLESv1_CM_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLESv2_nvidia.so.550.67 as /usr/lib/libGLESv2_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libGLX_nvidia.so.550.67 as /usr/lib/libGLX_nvidia.so.550.67
INFO[0000] Selecting /usr/lib/libcuda.so.550.67 as /usr/lib/libcuda.so.550.67
INFO[0000] Selecting /usr/lib/libcudadebugger.so.550.67 as /usr/lib/libcudadebugger.so.550.67
INFO[0000] Selecting /usr/lib/libnvcuvid.so.550.67 as /usr/lib/libnvcuvid.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-allocator.so.550.67 as /usr/lib/libnvidia-allocator.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-cfg.so.550.67 as /usr/lib/libnvidia-cfg.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-eglcore.so.550.67 as /usr/lib/libnvidia-eglcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-encode.so.550.67 as /usr/lib/libnvidia-encode.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-fbc.so.550.67 as /usr/lib/libnvidia-fbc.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glcore.so.550.67 as /usr/lib/libnvidia-glcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glsi.so.550.67 as /usr/lib/libnvidia-glsi.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-glvkspirv.so.550.67 as /usr/lib/libnvidia-glvkspirv.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-gpucomp.so.550.67 as /usr/lib/libnvidia-gpucomp.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ml.so.550.67 as /usr/lib/libnvidia-ml.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ngx.so.550.67 as /usr/lib/libnvidia-ngx.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-nvvm.so.550.67 as /usr/lib/libnvidia-nvvm.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-opticalflow.so.550.67 as /usr/lib/libnvidia-opticalflow.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-pkcs11-openssl3.so.550.67 as /usr/lib/libnvidia-pkcs11-openssl3.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-pkcs11.so.550.67 as /usr/lib/libnvidia-pkcs11.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-ptxjitcompiler.so.550.67 as /usr/lib/libnvidia-ptxjitcompiler.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-rtcore.so.550.67 as /usr/lib/libnvidia-rtcore.so.550.67
INFO[0000] Selecting /usr/lib/libnvidia-tls.so.550.67 as /usr/lib/libnvidia-tls.so.550.67
INFO[0000] Selecting /usr/lib/libnvoptix.so.550.67 as /usr/lib/libnvoptix.so.550.67
INFO[0000] Selecting /run/nvidia-persistenced/socket as /run/nvidia-persistenced/socket
WARN[0000] Could not locate /nvidia-fabricmanager/socket: pattern /nvidia-fabricmanager/socket not found
WARN[0000] Could not locate /tmp/nvidia-mps: pattern /tmp/nvidia-mps not found
INFO[0000] Selecting /lib/firmware/nvidia/550.67/gsp_ga10x.bin as /lib/firmware/nvidia/550.67/gsp_ga10x.bin
INFO[0000] Selecting /lib/firmware/nvidia/550.67/gsp_tu10x.bin as /lib/firmware/nvidia/550.67/gsp_tu10x.bin
INFO[0000] Selecting /usr/bin/nvidia-smi as /usr/bin/nvidia-smi
INFO[0000] Selecting /usr/bin/nvidia-debugdump as /usr/bin/nvidia-debugdump
INFO[0000] Selecting /usr/bin/nvidia-persistenced as /usr/bin/nvidia-persistenced
INFO[0000] Selecting /usr/bin/nvidia-cuda-mps-control as /usr/bin/nvidia-cuda-mps-control
INFO[0000] Selecting /usr/bin/nvidia-cuda-mps-server as /usr/bin/nvidia-cuda-mps-server
INFO[0000] Generated CDI spec with version 0.5.0
```
I don't see an indentation error in any of the two yaml files...

@klueska Podman always uses CDI, the `--gpus` flag works as an alias for `--device nvidia.com/gpu=`. See https://github.com/containers/podman/pull/21180

Reply:
Interesting -- that's new (and unfortunate). The `--gpus` flag is inconsistent and never should have been introduced in the first place. We'd prefer it to die a slow death than be added anew to other projects.

Reply:
> I don't see an indentation error in any of the two yaml files...

You are correct. That was strange, it was completely off the first time I loaded it. Must have been a GitHub presentation error. I just took the raw version this time.

Reply:
In terms of the logs, there are only minor differences:

```
$ diff logs-old.txt logs-new.txt
0a1,2
> $ ./build/nvidia-ctk cdi generate
> WARN[0000] Failed to locate nvidia-cdi-hook: pattern nvidia-cdi-hook not found
3,4c5,6
< INFO[0000] Selecting /dev/dri/card1 as /dev/dri/card1
< INFO[0000] Selecting /dev/dri/renderD128 as /dev/dri/renderD128
---
> WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--card; ignoring
> WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--render; ignoring
```

The first error is understood. It is looking for `nvidia-cdi-hook` but didn't find it. If you have a moment, and don't mind compiling it and putting it in a standard path, and seeing if that matters, that would help.

For the second set - the issue with `card` and `render` - I will have to dig deeper to try and understand.

Reply:
The differences in the yaml files at least partially are due to the card and render devices. Let's figure that out, and then see if the issue lingers.

Reply:
The successful log line is [here](https://github.com/NVIDIA/nvidia-container-toolkit/blob/fde099d25b98fc10886b49a7a3730f49c284b13e/internal/discover/mounts.go#L97), while the failure line is [here](https://github.com/NVIDIA/nvidia-container-toolkit/blob/fde099d25b98fc10886b49a7a3730f49c284b13e/pkg/nvcdi/full-gpu-nvml.go#L199). Not clear the flow from one to the other.

I wouldn't mind walking through this in a debugger. I don't assume there is any way to walk it through together @lahwaacz ?

Reply:
OK, the `WARN[0000] Failed to locate nvidia-cdi-hook: pattern nvidia-cdi-hook not found` is gone when I add the build directory to `$PATH`.

For the `Failed to evaluate symlink` warnings, it seems there is some variable expansion missing: `/dev/dri/by-path/pci--card` should be `/dev/dri/by-path/pci-0000:21:00.0-card` and similar for `render`.

Reply:
> OK, the WARN[0000] Failed to locate nvidia-cdi-hook: pattern nvidia-cdi-hook not found is gone when I add the build directory to $PATH.

One down, one to go (another town and one more show 🎶 ). Thanks, that helps.

> For the Failed to evaluate symlink warnings, it seems there is some variable expansion missing: /dev/dri/by-path/pci--card should be /dev/dri/by-path/pci-0000:21:00.0-card and similar for render

Ah, that makes sense. I will see if I can figure it out any further from here, and if not, we will look for alternate methods.

Reply:
I don't think that part of the code is changed, but I will keep trying.

> I don't assume there is any way to walk it through together @lahwaacz ?

If this is a possibility, let me know?

Reply:
@lahwaacz I added a commit with a whole bunch of debugging statements. Do you mind recompiling `nvidia-ctk`, and running generate again but with debug? `nvidia-ctk --debug cdi generate`? And posting logs?

Reply:
Here are the logs up to the two warnings:
```
DEBU[0000] Locating NVIDIA Container Toolkit CLI as nvidia-cdi-hook
DEBU[0000] Locating "nvidia-cdi-hook" in [/usr/local/sbin /usr/local/bin /usr/bin /opt/cuda/bin /opt/cuda/nsight_compute /opt/cuda/nsight_systems/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /home/lahwaacz/stuff/nvidia-container-toolkit/build /usr/sbin /sbin /bin]
DEBU[0000] Checking candidate '/home/lahwaacz/stuff/nvidia-container-toolkit/build/nvidia-cdi-hook'
DEBU[0000] Found 1 candidates; ignoring further candidates
DEBU[0000] Found nvidia-cdi-hook candidates: [/home/lahwaacz/stuff/nvidia-container-toolkit/build/nvidia-cdi-hook]
DEBU[0000] Using NVIDIA Container Toolkit CLI path nvidia-cdi-hook
DEBU[0000] Is WSL-based system? false: could not load DXCore library: libdxcore.so: cannot open shared object file: No such file or directory
DEBU[0000] Is NVML-based system? true: found NVML library
DEBU[0000] Is Tegra-based system? false: /sys/devices/soc0/family file not found
INFO[0000] Auto-detected mode as "nvml"
DEBU[0000] newFullGPUDiscoverer() devRoot / device &device.device{Device:nvml.nvmlDevice{Handle:(*nvml._Ctype_struct_nvmlDevice_st)(0x7a0371bfaf38)}, lib:(*device.devicelib)(0xc0000173c0), migProfiles:[]device.MigProfile(nil)}
DEBU[0000] newFullGPUDiscoverer() minor 0
DEBU[0000] newFullGPUDiscoverer() pciInfo nvml.PciInfo{BusIdLegacy:[16]int8{48, 48, 48, 48, 58, 50, 49, 58, 48, 48, 46, 48, 0, 0, 0, 0}, Domain:0x0, Bus:0x21, Device:0x0, PciDeviceId:0x1eb010de, PciSubSystemId:0x0, BusId:[32]int8{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}}
DEBU[0000] newFullGPUDiscoverer() pciBusID
DEBU[0000] Locating /dev/nvidia0
DEBU[0000] Locating "/dev/nvidia0" in [/ /dev]
DEBU[0000] Checking candidate '/dev/nvidia0'
DEBU[0000] Located /dev/nvidia0 as [/dev/nvidia0]
INFO[0000] Selecting /dev/nvidia0 as /dev/nvidia0
DEBU[0000] byPathHookDiscoverer.deviceNodeLinks() for nvcdi.byPathHookDiscoverer{logger:(*logrus.Logger)(0xc0000d4a00), devRoot:"/", nvidiaCDIHookPath:"/home/klinkovsky/stuff/nvidia-container-toolkit/build/nvidia-cdi-hook", pciBusID:"", deviceNodes:(*discover.charDevices)(0xc0001d6850)}
DEBU[0000] returning cached mounts
DEBU[0000] byPathHookDiscoverer.deviceNodeLinks() devices []discover.Device{discover.Device{HostPath:"/dev/nvidia0", Path:"/dev/nvidia0"}}
DEBU[0000] byPathHookDiscoverer.deviceNodeLinks() candidates [/dev/dri/by-path/pci--card /dev/dri/by-path/pci--render]
WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--card; ignoring
WARN[0000] Failed to evaluate symlink /dev/dri/by-path/pci--render; ignoring
```

Reply:
> DEBU[0000] newFullGPUDiscoverer() pciInfo nvml.PciInfo{BusIdLegacy:[16]int8{48, 48, 48, 48, 58, 50, 49, 58, 48, 48, 46, 48, 0, 0, 0, 0}, Domain:0x0, Bus:0x21, Device:0x0, PciDeviceId:0x1eb010de, PciSubSystemId:0x0, BusId:[32]int8{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}}
> DEBU[0000] newFullGPUDiscoverer() pciBusID

pciBudID is all all `0x00`. This part comes from go-nvlib, which wasn't changed. Does this have to do with how we compile it in? `make build` just does a `go build ./...`, so I am not sure. @klueska do you understand how this part works?

Reply:
Oh interesting. So I didn't introduce the bug. 

If you've compiled it from main, do you mind generating the CDI yaml from main, so we can compare? It doesn't get us all of the way, but it helps. 

Hopefully @klueska or one of the other maintainers will know what we need to do to complete the devices.

Reply:
@deitch CDI from main: https://gist.github.com/lahwaacz/ca4868a70ca106cb4a90580b6735595e

Reply:
Thanks @lahwaacz . Those are identical, except for the executable. That is a good sign.

It means that whatever introduced the issue with the `busId`, is in main, and was not introduced by this PR.

Reply:
I did a quick check of `main` vs `v1.15.0`, there is quite a bit in there that I do not understand. I think we are stuck until someone who gets how this works explains how to fix it in main (and we can apply here).